### PR TITLE
cleaned up PCR_mm_ValidateSequences

### PR DESCRIPTION
**Pull request by chr1sgrant**

### CODE REVIEW

**Comment by gogogadgetbioprocess** [comment]
depends on PR #220 of antha-lang/antha

**Comment by gogogadgetbioprocess** [comment]
@synthacebot test this please

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 1]
```diff
@@ -31,30 +34,25 @@ Parameters (
...
+	AnnealingTime Time
 	Finalextensiontime Time
 )
 
 // Data which is returned from this protocol, and data types
 Data (
+	Amplicon wtype.DNASequence
 	FwdPrimerSites []search.Thingfound
 	RevPrimerSites []search.Thingfound
+	Fwdprimermeltingtemp Temperature
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: FwdPrimerMeltingTemp

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 2]
```diff
@@ -31,30 +34,25 @@ Parameters (
...
+	Revprimermeltingtemp Temperature
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: RevPrimerMeltingTemp

File: defaultparameters/PCR_mmx_ValidateSequences.default.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (ddn0) [inline defaultparameters/PCR_mmx_ValidateSequences.default.json]: rename to end with bundle.json
and move to defaults/

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 4]
```diff
@@ -31,30 +34,25 @@ Parameters (
...
-	//Templatetype string  // e.g. colony, genomic, pure plasmid... will effect efficiency. We could get more sophisticated here later on... 
-	//FullTemplatesequence string // better to use Sid's type system here after proof of concept
-	//FullTemplatelength int	// clearly could be calculated from the sequence... Sid will have a method to do this already so check!
-	//TargetTemplatesequence string // better to use Sid's type system here after proof of concept
-	//TargetTemplatelengthinBP int
-	*/
 	// Reaction parameters: (could be a entered as thermocycle parameters type possibly?)
 	Numberofcycles int 
 	InitDenaturationtime Time
 	Denaturationtime Time
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: The case for the rest of them?

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 5]
```diff
@@ -90,17 +88,20 @@ Steps {
...
+	FwdPrimerSites = sequences.FindSeqsinSeqs(TemplateSequence.Sequence(), []string{FwdPrimerSeq.Sequence()})
+	
+	RevPrimerSites = sequences.FindSeqsinSeqs(TemplateSequence.Sequence(), []string{RevPrimerSeq.Sequence()})
 
 	
-	if len(FwdPrimerSites)==0 || len(RevPrimerSites)==0{
+	if len(FwdPrimerSites)!=1 || len(RevPrimerSites)!=1{
 	
 	errordescription := fmt.Sprint(
 						text.Print("FwdPrimerSitesfound:", fmt.Sprint(FwdPrimerSites)),
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Would it be more understandable to write: "No exact match found, (%d matches)" ?

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 6]
```diff
@@ -165,33 +190,62 @@ Steps {
...
+	if !found{
+		Errorf("No melting temp found for polymerase ",polymerase )
+	}
+
+	var err error
+
+	ExtensionTime, err = enzymes.CalculateExtensionTime(PCRPolymerase, Amplicon) 
+	
+	if err != nil {
+		Errorf("Can't calculate extensiontime of polymerase.",  err.Error())
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: "extension time" in the error message

**Comment by gogogadgetbioprocess** [comment]
all the requested changes are in now.

File: lib/PCR_mmx_ValidateSequences_.go [hunk 7]
```diff
@@ -127,32 +136,61 @@ func _PCR_ValidateSequencesSteps(_ctx context.Context, _input *PCR_ValidateSeque
...
+	// work out annealing temperature
+
+	_output.FwdPrimerMeltingTemp = oligos.BasicMeltingTemp(_input.FwdPrimerSeq)
+
+	_output.RevPrimerMeltingTemp = oligos.BasicMeltingTemp(_input.RevPrimerSeq)
+
+	// check which primer has the lowest melting temperature
+	if _output.FwdPrimerMeltingTemp.SIValue() < _output.RevPrimerMeltingTemp.SIValue() {
+		// start PCR 3 degress below lowest melting temp
+		_output.AnnealingTemp = wunit.NewTemperature(_output.FwdPrimerMeltingTemp.SIValue()-3.0, "C")
```
Comment (ddn0) [inline lib/PCR_mmx_ValidateSequences_.go]: Pull out constant.

```go
const (
  // Starting melting a few degrees below primer melting temperature
  primerMeltingAdjustment = -3.0
)
```

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 8]
```diff
@@ -165,33 +190,62 @@ Steps {
...
+	// work out annealing temperature
+	
+	FwdPrimerMeltingTemp = oligos.BasicMeltingTemp(FwdPrimerSeq)
+	
+	RevPrimerMeltingTemp = oligos.BasicMeltingTemp(RevPrimerSeq)
+
+	// check which primer has the lowest melting temperature
+	if FwdPrimerMeltingTemp.SIValue() < RevPrimerMeltingTemp.SIValue(){
+		// start PCR 3 degress below lowest melting temp
+		AnnealingTemp = wunit.NewTemperature(FwdPrimerMeltingTemp.SIValue()- 3.0,"C")
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Pull out constant.

```go
const (
  // Starting melting a few degrees below primer melting temperature
  primerMeltingAdjustment = -3.0
)
```

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an (hunk 5 above)
Comment (gogogadgetbioprocess) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: I don't personally think so

Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: The error condition is on `FwdPrimerSites` and `RevPrimerSites` yet the error message only mentions `FwdPrimerSites`.

**Comment by gogogadgetbioprocess** [comment]
depends on PR antha-lang/antha #253

**Comment by gogogadgetbioprocess** [comment]
@synthacebot test this please

**Comment by gogogadgetbioprocess** [comment]
The dependency PR antha-lang/antha #253  is merged in now so please can this be merged in now?

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 9]
```diff
@@ -165,33 +189,60 @@ Steps {
...
-	meltingTemp := polymeraseproperties["meltingtemp"]
+	var found bool
+
+	ExtensionTemp, found = polymeraseproperties["extensiontemp"]
+	if !found{
+		Errorf("No extension temp found for polymerase ",polymerase )
+	}
+	MeltingTemp, found = polymeraseproperties["meltingtemp"]
+	if !found{
+		Errorf("No melting temp found for polymerase ",polymerase )
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Errorf("... %s", polymerase)

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 10]
```diff
@@ -165,33 +189,60 @@ Steps {
 		Errorf("No Properties for", polymerase, "found.", "Valid options are:",strings.Join(validoptions,","))
 	}
 
-	extensionTemp := polymeraseproperties["extensiontemp"]
-	meltingTemp := polymeraseproperties["meltingtemp"]
+	var found bool
+
+	ExtensionTemp, found = polymeraseproperties["extensiontemp"]
+	if !found{
+		Errorf("No extension temp found for polymerase ",polymerase )
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Errorf("... %s", polymerase)

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 11]
```diff
@@ -165,33 +189,60 @@ Steps {
...
 
-	//extensiontime, _ := enzymes.CalculateExtensionTime(polymerase, targetsequence) 
+	// work out annealing temperature
+	
+	FwdPrimerMeltingTemp = oligos.BasicMeltingTemp(FwdPrimerSeq)
+	
+	RevPrimerMeltingTemp = oligos.BasicMeltingTemp(RevPrimerSeq)
+
+	// check which primer has the lowest melting temperature
+	if FwdPrimerMeltingTemp.SIValue() < RevPrimerMeltingTemp.SIValue(){
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]:
```go
lowest := math.Min(FwdPrimerMeltingTemp.SIValue(), RevPrimerMeltingTemp.SIValue())
AnnealingTemp = wunit.NewTemperature(lowest - AnnealingTempOffset.SIValue(), "C")
```

File: starter/AutoPCR/PCR.an [hunk 12]
```diff
@@ -52,7 +52,8 @@ Inputs (
 	PCRPolymerase *wtype.LHComponent
 	Buffer *wtype.LHComponent
 	Water *wtype.LHComponent
-	Template *wtype.LHComponent
+	Template *wtype.LHComponent // the name of the LHComponent; any LHComponent coming in from the parameters file must be a valid LHComponent so Antha knows how to handle it
```
Comment (ddn0) [inline starter/AutoPCR/PCR.an]: Please godoc parameters and camelCase

File: starter/AutoPCR/PCR.an [hunk 13]
```diff
@@ -69,14 +70,26 @@ Requirements {
...
+	// The LHComponent type has many properties and behaviours which you can call upon using a period
+	// For example, an LHComponent's name is stored as a field called CName. 
+	// We can change the name of the LHComponent Template to the string TemplateName like so
+	
 	Template.CName = TemplateName
+	// The Template and TemplateName variables are declared above and given a type
+	// Template is an LHComponent which is a physical input so it's declared in the Inputs section.
+	// TemplateName is a string so just data input; it is therefore delcared in the parameters section.
+	
+	// Now do the same for the primers
```
Comment (ddn0) [inline starter/AutoPCR/PCR.an]: This is for line 98: remove redundant parens ()
This is for line 102: var samples []*wtype.LHComponent
This is for line 111: Update Errorf to match other cases.
This is for line 119: if !Hotstart { ... }
This is for lines 128-130 (and similar): Correct indentation.

File: an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an [hunk 14]
```diff
@@ -70,9 +70,6 @@ Steps {
 	// initialise map
```
Comment (ddn0) [inline an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an]: Line 1-48: camelCase, godoc parameters
Line 67: unnecessary
Line 68: var volumes []wunit.Volume
Line 69: var welllocations []string

File: an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an [hunk 15]
```diff
@@ -70,9 +70,6 @@ Steps {
...
 	ReactionMap = make(map[string]*wtype.LHComponent)
 
-
-	// range through the Reaction to template map
-
 	additives := make([]*wtype.LHComponent,0)
```
Comment (ddn0) [inline an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an]: var additives []*wtype.LHComponent
var additivevolumes []wunit.Volume

File: an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an [hunk 16]
```diff
@@ -92,7 +89,7 @@ Steps {
 		additivevolumes = append(additivevolumes,volume)
 	}
 
- 	
+ 	// range through the Reaction to template map
 	for reactionname, templatename := range Reactiontotemplate {
 	
 	// use counter to find next available well position in plate
```
Comment (ddn0) [inline an/AnthaAcademy/Lesson0_Examples/AutoPCR/AutoPCR.an]: Line 99: whitespace formatting
Lines 103: ditto
Line 159: Errorf(errormessage)

File: an/AnthaAcademy/Lesson0_Examples/AutoPCR/PCR.an [hunk 17]
```diff
@@ -102,7 +104,7 @@ Steps {
 	samples = append(samples, dntpSample)
 	
 	if len(Additives)!=len(AdditiveVols){
-		Errorf("Bad things are going to happen if you have different numbers of additives and additivevolumes")
+		Errorf("Additives and AdditiveVols need to contain the same number of entries, check this please")
 	}
 	
 	for i := range Additives {
```
Comment (ddn0) [inline an/AnthaAcademy/Lesson0_Examples/AutoPCR/PCR.an]: Line 115: if !Hotstart {...}
Line 117: formatting
Line 123-126: formatting
Line ***: formatting, initializing slices as `var x []type`
Line 219: delete block commented code and dead code

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 18]
```diff
@@ -87,28 +86,53 @@ Setup {
...
+	
+	
+	// work out what the pcr product will be
+	Amplicon = oligos.DNAregion(TemplateSequence,startposition,endposition)
+	
+	
+	
 	// Make a mastermix
 
 	mmxSample := mixer.Sample(MasterMix, MasterMixVolume)
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Lines 151-***: format

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 19]
```diff
@@ -165,33 +189,60 @@ Steps {
 		Errorf("No Properties for", polymerase, "found.", "Valid options are:",strings.Join(validoptions,","))
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Lines 184: var xxx []string
Lines 186: formatting

File: an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an [hunk 20]
```diff
@@ -165,33 +189,60 @@ Steps {
...
-	r1 = Incubate(r1, extensionTemp, Extensiontime, false)
+	r1 = Incubate(r1, ExtensionTemp, ExtensionTime, false)
 		
 	}
 		// Final Extension
-	r1 = Incubate(r1, extensionTemp, Finalextensiontime, false)
+	r1 = Incubate(r1, ExtensionTemp, FinalExtensionTime, false)
 	
 
 	// all done
```
Comment (ddn0) [inline an/Liquid_handling/PCR/PCR_mmx_ValidateSequences.an]: Lines 265: delete commented code

**Pull request merged**